### PR TITLE
(#40) Autosuggest Starts With Logic Update

### LINKED ISSUE #40: Move Autosuggest "Starts with" business logic to the frontend
### Issue description
There was a business logic tied to the "Starts with" functionality of the Autosuggest feature that was in dictionary api which has now been removed to allow for better flexibility of the api. This business logic would now be moved to the frontend since it's only peculiar to our use case. 

> **ESTIMATE** 5

### What's the expected change?
- Since the "Begins" logic has been removed from the API, the frontend needs to implement this logic by making
"Starts with" API requests in this order:
1. Starts with DrugTerm
2. Starts with DrugAlias if step 1 above returns less than 10 results, merge response with that from step 1
3. Contains Drug Alias if result of step 2 above returns less than 10 results, merge response with that from step 2

### What's the current functionality?
- The frontend makes a "Starts with" fetch request to the dictionary API's Autosuggest endpoint that performs a "Begins" search and then, if the maximum results limit hadn't been reached, perform a "Contains" search as well and append those results before returning a response to the frontend.

### What's the updated acceptance criteria?
```gherkin
  Scenario: User is able to see autosuggested items when typing 3 or more characters into the search box and has selected "Starts with"
    Given the user navigates to "def/bevacizumab"
     When user selects "Starts with" option
     When user clicks on the search bar
     When user types "bin" in the search bar
     Then autosuggest appears with correct options
      | options                                                                                                    | 
      | binimetinib                                                                                                | 
      | bintrafusp alfa                                                                                            | 
      | BinD 19                                                                                                    | 
      | BIND-014                                                                                                   | 
      | BinD19                                                                                                     | 
      | Binosto                                                                                                    | 
      | 1,1',6,6',7,7'-Hexahydroxy-3,3'-dimethyl-5,5'-bis(1-methylethyl)[2,2'-binaphthalene]-8,8'-dicarboxaldehyde | 
      | 1,1',6,6',7,7'-Hexahydroxy-5,5'-diisopropyl-3-3'-dimethyl[2,2'-binaphthalene]-8,8'-dicarboxaldehyde        | 
      | autologous anti-CD19 CAR TCR-zeta/4-1BB-transduced T cells BinD19                                          | 
      | CBC 80/20 binder ABX-464                                                                                   | 

  Scenario: User is able to see autosuggested items when typing 3 or more characters into the search box and has selected "Contains"
    Given the user navigates to "def/bevacizumab"
     When user selects "Contains" option
     When user clicks on the search bar
     When user types "bev" in the search bar
     Then autosuggest appears with correct options
      | options                                                         | 
      | carboplatin-Taxol-bevacizumab regimen                           | 
      | carrot/Ji-Lin ginseng/licorice root/tangerine peel soy beverage | 
      | fermented soybean protein beverage                              | 
      | FOLFIRI-bevacizumab regimen                                     | 
      | indium In 111 bevacizumab                                       | 
      | 111-indium labeled bevacizumab                                  | 
      | carboplatin-paclitaxel-bevacizumab regimen                      | 
      | carrot/Jilin ginseng/licorice root/tangerine peel soy beverage  | 
      | electro-kinetically altered beverage                            | 
      | folinic acid-fluorouracil-irinotecan-bevacizumab regimen        | 

```

### Additional details / screenshot
- Screenshot

### Related Tickets
* Original Story Ticket #9999
* Original Story Ticket #9999

#### Technical implementation for Starts with:
API needs to make three separate queries. 

1. Starts with DrugTerm
2. Starts with DrugAlias
3. Contains Drug Alias

If first query returns >= 10 results, the second and third queries are not needed. If first query returns < 10 results, the second query's result is appended after first query's result (the result of the third query is discarded if the merged results of the first and second queries is >= 10). Results of the third query is added if 1st and 2nd merged results is < 10.

Contains: is just a pure contains

### PR DESCRIPTION
**Pull request by seyilonge-nci**
Closes #40

* Added custom hook for autosuggest fetching
* Updated api to point to prod
* Updated unit tests
* Put back encodedURIComponent(), added 'Contains' search type to early exit
* Encoded expected test search string
* Refactored autoSuggestQuery hook, updated autosuggest input box display logic, added mock files for new call
* Applied prettier formatting

### CODE REVIEW

**Comment by seyilonge-nci** [comment]
The auto-suggest options displayed when a search text is entered into the input box is populated by three separate api calls that are merged into one result to display a maximum number of 10 results as suggestions, with matching text bolded in the suggestion options (case insensitive bold highlighting).

Please note that both in current production and this implementation, results returned to be displayed are not sorted. This in essence means that for the same text entered, as long as options displayed are 10 results (which is what the displayed results are limited to for the options), there could be more than 10 matching results returned but not all displayed. Since results are not sorted, displayed results could vary per hit.

For testing please consider the following:
* For "Starts with", as long as displayed options is less than 10 items. Options displayed should match prod. NOTE: Did find a peculiar scenario for "acetamido" where this is an exclusion.
* For "Starts with", if displayed options equal 10 items. Options displayed most likely wouldn't be an exact match as prod, and could actually vary between various hit instances for the same search text on react-app-dev.

**Comment by blilianyu** [comment]
@seyilonge-nci This looks good to us! 
cc: @mworrest @VictoriaSunNIH @lburack

**Review by karlikpj — APPROVED** [review]

File: src/components/molecules/search/Search.jsx [hunk 1]
```diff
@@ -137,35 +142,34 @@ const Search = ({ autoSuggestLimit = 10 }) => {
...
+						role="listbox"
+						data-testid="tid-auto-suggest-options">
+						{searchText.length >= 3 ? (
+							!autoSuggest.loading && autoSuggest.payload?.length ? (
+								children
+							) : (
+								<div className="ncids-autocomplete__menu-item">
+									{autoSuggest.loading
+										? 'Loading results...'
+										: 'No results found'}
```
Comment (kate-mashkina) [inline src/components/molecules/search/Search.jsx:163]: We have removed 'no results found' helper text previously with bug fix - this reintroduces it. Could you replace it with an empty tag

Comment (seyilonge-nci) [reply src/components/molecules/search/Search.jsx]: Thanks for the catch, missed in rebase. Fixed!

**Comment by kate-mashkina** [comment]
Please find QA testing notes here https://lion.box.com/s/fv80k561446w4loytkt6lds1l0qw19il

**Comment by alinai** [comment]
Some of the issues noticed during testing on the react-dev:

- Starts With 3-0 returns no results.
All three query return empty arrays, but prod returns one term - prettyUrl name testolactone. Now API response for that term HAS an alias that contains 3-0 (see response link below) 
https://webapis-dev.cancer.gov/drugdictionary/v1/Drugs/testolactone

- When searched for Starts With 'avem' autosuggest displays duplicate items, because Api returns duplicate aliases- https://webapis-dev.cancer.gov/drugdictionary/v1/Autosuggest?searchText=avem&matchType=Begins&size=20&includeResourceTypes=DrugAlias
Perhaps we can place a filter to only display distinct results ? in response you notice it is a same term , just US Brand Name and Foreign brand name got identical values
 
- searching contains s-m does not display any results, but s-ma returns one

- horizontal scroll appears on typing contains 'pyr' or 'ade', starts with for '11a'
image

- Starts with - type lymphocyte and add a space - react shows 2 results, whereas prod shows 10 (and more probably). The third call in react(contains aliases) also returns bunch of results, but they are no displayed - is this related to https://github.com/NCIOCPL/drug-dictionary-app/issues/39 ?

- Starts with for "zum" , "IL2" , "223" react displays less results than prod and for "ascorbic acid" and "melphalan" react displays more results than prod even if number of results in the autosuggest <10  

- Contains- For Contains most of the time React displays more results than Prod even if number of results is less than 10. Example: bgb, pend, tem

- Contains- Type "I 13" and see that 8 results are displayed on react, but API for contains call returns 10 results. 

- Contains - Typing "12/eg" on Prod displays autosuggest terms but React does not. React has this term "4H11-28z/fIL-12/EGFRt-expressing autologous T lymphocytes"

- Contains - Typing "28z" returns whole different sets of results on react compared to Prod though number of results <10

- Contains- Typing "ycd/" on Prod displays autosuggest terms but React does not. React has this term "Ad5-yCD/mutTKSR39rep-hIL12" and "Ad5-yCD/mutTK(SR39)rep-ADP"

- Starts with for '11a' gives more results on react than prod

- Contains- Typing "35(27L)"  and "1:26" on Prod displays autosuggest terms but React does not. Even though you'd think it should as there is a term  "MART-1:26-35(27L) peptide vaccine" 

- Contains - "350" gives different results than Prod

- Contains - "398" gives 3 results on React but none on prod

- Contains - "3.A" gives one result in React and 2 on Prod and no common result
-

**Comment by seyilonge-nci** [comment]
> Some of the issues noticed during testing on the react-dev:
> 
> * Starts With 3-0 returns no results.
>   All three query return empty arrays, but prod returns one term - prettyUrl name testolactone. Now API response for that term HAS an alias that contains 3-0 (see response link below) 
>   https://webapis-dev.cancer.gov/drugdictionary/v1/Drugs/testolactone

Fetch api queries listed below:
https://webapis.cancer.gov/drugdictionary/v1/Autosuggest?searchText=3-0&matchType=Begins&includeResourceTypes=DrugTerm&size=10

https://webapis.cancer.gov/drugdictionary/v1/Autosuggest?searchText=3-0&matchType=Begins&includeResourceTypes=DrugAlias&size=10

https://webapis.cancer.gov/drugdictionary/v1/Autosuggest?searchText=3-0&matchType=Contains&includeResourceTypes=DrugAlias&size=10

These do not return any results. @bryanpizzillo As far as I can tell the frontend is making the correct calls in order with the right api parameters @blairl-nih @sarinapadilla 

> * When searched for Starts With 'avem' autosuggest displays duplicate items, because Api returns duplicate aliases- https://webapis-dev.cancer.gov/drugdictionary/v1/Autosuggest?searchText=avem&matchType=Begins&size=20&includeResourceTypes=DrugAlias
>   Perhaps we can place a filter to only display distinct results ? in response you notice it is a same term , just US Brand Name and Foreign brand name got identical values

[ ] A filter will need to be applied on the frontend for distinct results


> * searching contains s-m does not display any results, but s-ma returns one

@bryanpizzillo For this, since we are doing a "Contains" with term name and alias as against current production that does just term name alone. Without the result size limitation for the call to the api (https://webapis.cancer.gov/drugdictionary/v1/Autosuggest?searchText=s-m&matchType=Contains&includeResourceTypes=DrugTerm), it returns more than 10 results with the expected result as in prod (which is the situation we had mentioned earlier, but in this case yields a less favorable result). I can have the "Contains" just run against the term name and not alias so we can have this match prod. What do you think? 

> * horizontal scroll appears on typing contains 'pyr' or 'ade', starts with for '11a'
>   image

[ ] Already applied a fix for this. Would be in the next push

> * Starts with - type lymphocyte and add a space - react shows 2 results, whereas prod shows 10 (and more probably). The third call in react(contains aliases) also returns bunch of results, but they are no displayed - is this related to #39 ?

Yes, this is related to issue #39 and should be resolved once that fix is in.

> * Starts with for "zum" , "IL2" , "223" react displays less results than prod and for "ascorbic acid" and "melphalan" react displays more results than prod even if number of results in the autosuggest <10

Probable api issue

> * Contains- For Contains most of the time React displays more results than Prod even if number of results is less than 10. Example: bgb, pend, tem

[ ] Would be fixed once "Contains" uses just drug term

> * Contains- Type "I 13" and see that 8 results are displayed on react, but API for contains call returns 10 results.
Would be fixed once "Contains" uses just drug term

> * Contains - Typing "12/eg" on Prod displays autosuggest terms but React does not. React has this term "4H11-28z/fIL-12/EGFRt-expressing autologous T lymphocytes"
Possible api issue

> * Contains - Typing "28z" returns whole different sets of results on react compared to Prod though number of results <10
> * Contains- Typing "ycd/" on Prod displays autosuggest terms but React does not. React has this term "Ad5-yCD/mutTKSR39rep-hIL12" and "Ad5-yCD/mutTK(SR39)rep-ADP"
> * Starts with for '11a' gives more results on react than prod
> * Contains- Typing "35(27L)"  and "1:26" on Prod displays autosuggest terms but React does not. Even though you'd think it should as there is a term  "MART-1:26-35(27L) peptide vaccine"
> * Contains - "350" gives different results than Prod
> * Contains - "398" gives 3 results on React but none on prod
> * Contains - "3.A" gives one result in React and 2 on Prod and no common result

**Comment by seyilonge-nci** [comment]
@kate-mashkina and @alinai 
All issues brought up have been addressed. Please refer to updated notes above.

**Comment by alinai** [comment]
Per @seyilonge-nci , the following is actually as a result of issue https://github.com/NCIOCPL/drug-dictionary-app/issues/39

"Contains- Type "I 13" and see that 8 results are displayed on react, but API for contains call returns 10 results.
Would be fixed once "Contains" uses just drug term"

**Comment by alinai** [comment]
The horizontal scroll does not appear anymore on typing "pyr" for contains type match, but the autosuggest term is appearing on the page (out of the autosuggest drop-down window).

image
image

**Comment by alinai** [comment]
For s-m contains type match, we get 10 results for drugterm and 10 results for drugalias in API calls but see one term on the autosuggest. Confirmed with @seyilonge-nci that it is related to #39.

**Comment by seyilonge-nci** [comment]
> The horizontal scroll does not appear anymore on typing "pyr" for contains type match, but the autosuggest term is appearing on the page (out of the autosuggest drop-down window).
> 
> image
> image

Fixed!

**Comment by alinai** [comment]
Confirmed with @seyilonge-nci  that the following is partially related to the api issue (https://github.com/NCIOCPL/drug-dictionary-api/issues/42) opened and flagged as not an issue, and also the fact that we now match against aliases,

- Contains - Typing "28z" returns whole different sets of results on react compared to Prod though number of results <10
- Contains- Typing "ycd/" on Prod displays autosuggest terms but React does not. React has this term "Ad5-yCD/mutTKSR39rep-hIL12" and "Ad5-yCD/mutTK(SR39)rep-ADP"
- Starts with for '11a' gives more results on react than prod

**Comment by alinai** [comment]
Still seeing the autosuggest term appearing on the page (out of the autosuggest drop-down window) on typing "pyr" for contains type match on IE at desktop breakpoint (100% zoom).

image

**Comment by alinai** [comment]
For the following issue reported above:
Contains- Typing "35(27L)"  and "1:26" on Prod displays autosuggest terms but React does not. Even though you'd think it 
should as there is a term  "MART-1:26-35(27L) peptide vaccine"
- [x] Same api issue: NCIOCPL/drug-dictionary-api#42
Contains - "350" gives different results than Prod
- [x] Same api issue: NCIOCPL/drug-dictionary-api#42
Contains - "398" gives 3 results on React but none on prod
- [x] React returns results for Contains Alias
Contains - "3.A" gives one result in React and 2 on Prod and no common result
- [x] related to issue #39

Noticed that sometimes at the front end we get fewer results (example text: recep)  or no results (example text: 2/3) the first time we type in a text and then get more autosuggest results or results the next time we type in. This is happening for both Starts with and Contains. It happens only once in a while.

**Comment by alinai** [comment]
Horizontal scroll is back on IE

image

**Review by kate-mashkina — APPROVED** [review]

**Comment by alinai** [comment]
Entered https://github.com/NCIOCPL/drug-dictionary-app/issues/80 for the IE issue reported above.

**Review by arcepaul — APPROVED** [review]

**Pull request merged**